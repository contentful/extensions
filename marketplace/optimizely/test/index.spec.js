/* global global */
import React from 'react';
import { cleanup, render } from '@testing-library/react';

import App from '../src';
global.window.close = () => {};
global.window.encodeURIComponent = x => x;
global.window.addEventListener = jest.fn();

global.window.localStorage = {
  getItem: () => {},
  setItem: () => {}
};

const LOCATIONS = {
  LOCATION_ENTRY_FIELD: 'entry-field',
  LOCATION_ENTRY_FIELD_SIDEBAR: 'entry-field-sidebar',
  LOCATION_ENTRY_SIDEBAR: 'entry-sidebar',
  LOCATION_DIALOG: 'dialog',
  LOCATION_ENTRY_EDITOR: 'entry-editor',
  LOCATION_PAGE: 'page',
  LOCATION_APP: 'app'
};

let LOCATION = '';
let PROJECT_ID = '';
let VALID_FIELDS = false;

function mockSdk() {
  return {
    parameters: {
      installation: {
        optimizelyProjectId: PROJECT_ID
      }
    },
    location: {
      is: jest.fn(l => {
        return l === LOCATION;
      })
    },
    window: {
      startAutoResizer: () => {},
      stopAutoResizer: () => {}
    },
    ids: {},
    space: {},
    locales: {},
    entry: {
      fields: {
        experimentId: {
          getValue: jest.fn(() => 'exp123'),
          onValueChanged: jest.fn(() => jest.fn())
        },
        meta: {
          getValue: jest.fn(),
          onValueChanged: jest.fn(() => jest.fn())
        },
        variations: {
          getValue: jest.fn(),
          onValueChanged: jest.fn(() => jest.fn())
        }
      }
    },
    contentType: {
      sys: {
        space: {
          sys: {
            type: 'Link',
            linkType: 'Space',
            id: 'cyu19ucaypb9'
          }
        },
        id: 'variationContainer',
        type: 'ContentType',
        createdAt: '2019-05-24T07:45:48.863Z',
        updatedAt: '2019-05-30T04:28:43.488Z',
        environment: {
          sys: {
            id: 'master',
            type: 'Link',
            linkType: 'Environment'
          }
        },
        revision: 3
      },
      name: 'Variation Container',
      description: null,
      displayField: 'experimentTitle',
      fields: [
        {
          id: 'experimentTitle',
          name: 'Experiment title',
          type: 'Symbol',
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false
        },
        {
          id: 'experimentId',
          name: 'Experiment ID',
          type: 'Symbol',
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false
        },
        {
          id: 'meta',
          name: 'Meta',
          type: 'Object',
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false
        },
        {
          id: 'variations',
          name: 'Variations',
          type: 'Array',
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false,
          items: {
            type: 'Link',
            validations: [],
            linkType: 'Entry'
          }
        }
      ].concat(
        VALID_FIELDS
          ? {
              id: 'experimentKey',
              name: 'Experiment key',
              type: 'Symbol',
              localized: false,
              required: false,
              validations: [],
              disabled: false,
              omitted: false
            }
          : []
      )
    }
  };
}

describe('Optimizely App', () => {
  afterEach(() => {
    cleanup();

    LOCATION = '';
    PROJECT_ID = '';
    VALID_FIELDS = false;
  });

  it('should render the missing project on the sidebar', () => {
    LOCATION = LOCATIONS.LOCATION_ENTRY_SIDEBAR;
    const sdk = mockSdk();
    const { getByTestId } = render(<App sdk={sdk} />);

    expect(getByTestId('missing-project')).toMatchSnapshot();
  });

  it('should render the sidebar', () => {
    LOCATION = LOCATIONS.LOCATION_ENTRY_SIDEBAR;
    PROJECT_ID = '123';
    const sdk = mockSdk();
    const { getByTestId } = render(<App sdk={sdk} />);

    expect(getByTestId('sidebar')).toMatchSnapshot();
  });

  it('should render the incorrect type message', () => {
    LOCATION = LOCATIONS.LOCATION_ENTRY_EDITOR;
    PROJECT_ID = '123';
    const sdk = mockSdk();

    const { getByTestId } = render(<App sdk={sdk} />);
    expect(getByTestId('incorrect-type')).toMatchSnapshot();
  });

  it('should render the editor page', () => {
    LOCATION = LOCATIONS.LOCATION_ENTRY_EDITOR;
    PROJECT_ID = '123';
    VALID_FIELDS = true;
    const sdk = mockSdk();

    const { getByTestId } = render(<App sdk={sdk} />);
    expect(getByTestId('editor-page')).toMatchSnapshot();
  });
});
