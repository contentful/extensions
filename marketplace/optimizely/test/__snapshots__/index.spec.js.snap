// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optimizely App should render the editor page 1`] = `
<div
  class="css-175t7kh"
  data-test-id="editor-page"
>
  <div
    class="css-3p38p2"
  >
    <div
      class="css-xieh1a"
    >
      <svg
        class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
      <span>
        Select experiment
      </span>
    </div>
    <svg
      class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
      data-test-id="cf-ui-icon"
      height="18"
      viewBox="0 0 24 24"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
    <div
      class="css-xieh1a"
    >
      <svg
        class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
      <span>
        Add content
      </span>
    </div>
    <svg
      class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
      data-test-id="cf-ui-icon"
      height="18"
      viewBox="0 0 24 24"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
    <div
      class="css-xieh1a"
    >
      <svg
        class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
      <span>
        Publish variations
      </span>
    </div>
    <svg
      class="Icon__Icon___38Epv css-13af3fo Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
      data-test-id="cf-ui-icon"
      height="18"
      viewBox="0 0 24 24"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
    <div
      class="css-xieh1a"
    >
      <svg
        class="Icon__Icon___38Epv css-uoder Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
        />
      </svg>
      <span>
        Start experiment
      </span>
    </div>
  </div>
  <hr
    class="css-14380qe"
  />
  <h2
    class="Heading__Heading___83D3K css-1fobf8d"
    data-test-id="cf-ui-heading"
  >
    Referenced in:
  </h2>
  <div
    class="css-1y5g446"
  >
    <svg
      aria-label="Loading component..."
      class="SkeletonContainer__SkeletonContainer___23jiu"
      data-test-id="cf-ui-skeleton-form"
      height="30px"
      preserveAspectRatio="xMidYMid meet"
      role="img"
      width="100%"
    >
      <title>
        Loading component...
      </title>
      <rect
        clip-path="url(#references-section)"
        height="100%"
        style="fill: url(#cf-ui-skeleton-clip-gradient);"
        width="100%"
        x="0"
        y="0"
      />
      <defs>
        <clippath
          id="references-section"
        >
          <rect
            height="16"
            rx="0"
            ry="0"
            width="100%"
            x="0"
            y="0"
          />
        </clippath>
        <lineargradient
          id="cf-ui-skeleton-clip-gradient"
        >
          <stop
            offset="0%"
            stop-color="#e5ebed"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              dur="2s"
              id="animation1"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
          <stop
            offset="50%"
            stop-color="#f7f9fa"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              begin="animation1.begin+0.25s"
              dur="2s"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
          <stop
            offset="100%"
            stop-color="#e5ebed"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              begin="animation1.begin+0.5s"
              dur="2s"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
        </lineargradient>
      </defs>
    </svg>
  </div>
  <hr
    class="css-14380qe"
  />
  <h2
    class="Heading__Heading___83D3K css-1fobf8d"
    data-test-id="cf-ui-heading"
  >
    Experiment:
  </h2>
  <div
    class="SelectField__SelectField___kbQlf"
    data-test-id="cf-ui-select-field"
  >
    <div
      class="SelectField__SelectField__label-wrapper___3jGwo"
    >
      <label
        class="FormLabel__FormLabel___3d6zQ"
        data-test-id="cf-ui-form-label"
        for="experiment"
      >
        Optimizely experiment
        <span
          class="FormLabel__FormLabel__required-text___3mvdm"
        >
          (
          required
          )
        </span>
      </label>
    </div>
    <div
      class="Select__Select__wrapper___2pBIz Select__Select--large___1PgJW"
    >
      <select
        aria-label="experiment"
        class="Select__Select___31Z46 a11y__focus-border--default___60AXp Select__Select--disabled___3n58N"
        data-test-id="cf-ui-select"
        disabled=""
        id="experiment"
        name="experiment"
        required=""
      >
        <option
          data-test-id="cf-ui-select-option"
          value="-1"
        >
          Fetching experiments...
        </option>
      </select>
      <svg
        class="Icon__Icon___38Epv Select__Select__icon___boRAZ Icon__Icon--small___1yGZK Icon__Icon--muted___3egnD"
        data-test-id="cf-ui-icon"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
  <hr
    class="css-14380qe"
  />
  <h2
    class="Heading__Heading___83D3K"
    data-test-id="cf-ui-heading"
  >
    Variations:
  </h2>
  <div
    class="css-1pfvl5o"
  >
    <svg
      aria-label="Loading component..."
      class="SkeletonContainer__SkeletonContainer___23jiu"
      data-test-id="cf-ui-skeleton-form"
      height="100"
      preserveAspectRatio="xMidYMid meet"
      role="img"
      width="100%"
    >
      <title>
        Loading component...
      </title>
      <rect
        clip-path="url(#variations-section)"
        height="100%"
        style="fill: url(#cf-ui-skeleton-clip-gradient);"
        width="100%"
        x="0"
        y="0"
      />
      <defs>
        <clippath
          id="variations-section"
        >
          <rect
            height="16"
            rx="0"
            ry="0"
            width="100%"
            x="0"
            y="20"
          />
          <rect
            height="16"
            rx="0"
            ry="0"
            width="100%"
            x="0"
            y="44"
          />
          <rect
            height="16"
            rx="0"
            ry="0"
            width="80%"
            x="0"
            y="68"
          />
        </clippath>
        <lineargradient
          id="cf-ui-skeleton-clip-gradient"
        >
          <stop
            offset="0%"
            stop-color="#e5ebed"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              dur="2s"
              id="animation1"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
          <stop
            offset="50%"
            stop-color="#f7f9fa"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              begin="animation1.begin+0.25s"
              dur="2s"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
          <stop
            offset="100%"
            stop-color="#e5ebed"
            stop-opacity="1"
          >
            <animate
              attributeName="stop-color"
              begin="animation1.begin+0.5s"
              dur="2s"
              repeatCount="indefinite"
              values="#e5ebed; #f7f9fa; #e5ebed"
            />
          </stop>
        </lineargradient>
      </defs>
    </svg>
  </div>
</div>
`;

exports[`Optimizely App should render the incorrect type message 1`] = `
<div
  class="css-hvj582"
  data-test-id="incorrect-type"
>
  <div
    class="Note__Note___2eSKN Note__Note--negative___3fkwL"
    data-test-id="cf-ui-note"
  >
    <div
      class="Note__Note__icon___20RqC"
    >
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--negative___1dled"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
    <div>
      <div
        class="Note__Note__title___2nwpx"
      >
        Extension "undefined" cannot be assigned to this content type
      </div>
      <div>
        <div>
          <strong>
            Required:
          </strong>
           
          experimentTitle (Symbol), experimentId (Symbol), experimentKey (Symbol), meta (Object), variations (Array)
        </div>
        <div>
          <strong>
            Missing:
          </strong>
           
          experimentKey (Symbol)
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Optimizely App should render the missing project on the sidebar 1`] = `
<div
  class="css-hvj582"
  data-test-id="missing-project"
>
  <div
    class="Note__Note___2eSKN Note__Note--negative___3fkwL"
    data-test-id="cf-ui-note"
  >
    <div
      class="Note__Note__icon___20RqC"
    >
      <svg
        class="Icon__Icon___38Epv Icon__Icon--small___1yGZK Icon__Icon--negative___1dled"
        data-test-id="cf-ui-icon"
        height="18"
        viewBox="0 0 24 24"
        width="18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
    <div>
      <div>
        Please, set 
        <strong>
          Optimizely Project ID
        </strong>
         in Extension settings.
      </div>
    </div>
  </div>
</div>
`;

exports[`Optimizely App should render the sidebar 1`] = `
<div
  data-test-id="sidebar"
>
  <a
    class="Button__Button___1ZfFj a11y__focus-border--default___60AXp css-82qlwu Button__Button--primary___JImeO Button__Button--full-width___3Fmpo"
    data-test-id="view-experiment"
    href="https://app.optimizely.com/v2/projects/123/experiments/exp123/variations"
    target="_blank"
    type="button"
  >
    <span
      class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
      tabindex="-1"
    >
      <span
        class="Button__Button__label___3tcOj"
      >
        View in Optimizely
      </span>
    </span>
  </a>
  <a
    class="Button__Button___1ZfFj a11y__focus-border--default___60AXp css-82qlwu Button__Button--muted___2Wair Button__Button--full-width___3Fmpo"
    data-test-id="view-all"
    href="https://app.optimizely.com/v2/projects/123/experiments"
    target="_blank"
    type="button"
  >
    <span
      class="TabFocusTrap__TabFocusTrap___39Vty Button__Button__inner-wrapper___3qrNC"
      tabindex="-1"
    >
      <span
        class="Button__Button__label___3tcOj"
      >
        View all experiments
      </span>
    </span>
  </a>
</div>
`;
